os/20201103-full-disk-raid.md

```diff
@@ -0,0 +1,208 @@
+
+## BIOS Boot partition
+
+The Dracut glue will save an image of the BIOS boot partition whenever the config specifies a partition with the BIOS boot type GUID, and will restore that image to _every_ such partition.  In addition, the same logic will save the boot sector (the first 440 bytes of the disk) and restore it to every disk where we write a BIOS Boot partition.
+
+The BIOS Boot partition is also not updated at runtime, so RAID is not necessary here either.  As a result, when updating the bootloader, bootupd will need to find each BIOS Boot type GUID and update both the partition and the corresponding boot sector.  Again this assumes that the entire system belongs to CoreOS.
+
+For simplicity, we're copying the BIOS Boot partition and boot sector bit-for-bit, rather than rerunning `grub-install` from the initrd or hand-modifying the boot sector.  Since the boot sector hardcodes the offset of the BIOS Boot partition, the latter must be recreated at the same offset as the original partition.  To avoid hardcoding an awkward constant in FCCT, we'll move the BIOS boot partition from its current offset of 512 MiB to the beginning of the disk (offset 1 MiB) in new images.  The Dracut glue will detect relocation of the BIOS boot partition and fail the boot.
+
+There's no BIOS boot partition in 4Kn images.  We don't know in advance which disks will be 4Kn, so the Ignition config will always create a BIOS boot partition on x86_64, but the Dracut glue will automatically skip copying it (and the boot sector) if there's no source partition to copy from.
+
+## PReP partition
+
+For ppc64le, we'll create independent replicas of the PReP partition, similar to the BIOS boot partition.  We'll need to confirm with multi-arch folks that this will work.
+
+## Ignition changes
+
+The Dracut glue can detect the requisite Ignition config stanzas using `jq`, but that's brittle and we're already doing a lot of it in the initrd.  Ideally Ignition will provide command-line arguments for executing relevant (hard-coded) queries against the cached Ignition config.
+
+## Architecture support
+
+We'll need to get feedback from the multi-arch team for aarch64 and ppc64le.  ppc64le will need an alternate `layout` definition that includes the PReP partition.
+
+For s390x, Ignition doesn't yet support DASD partitioning, and we'd need provisions for rerunning zipl from the initramfs, if RAID support is even possible at all.  Each DASD is connected via a particular bus identifier, which might need to be encoded into each respective copy of the bootloader.  We may need to declare full-disk RAID out of scope for s390x, at least for 4.7.
+
+## Configuring boot disk RAID in OCP
+
+Since FCCs aren't yet used directly in MachineConfigs and FCCT is not shipped with OCP, we'll need a workaround to allow FCC sugar to be used in 4.7.  The MCO could:
```

Comment:
xref https://github.com/openshift/enhancements/pull/525

I'm uncertain though that the MCO is required here; I think in some cases this may need to be per-machine state and actually be part of the pointer config, which argues for having `oc ignition fcct`/`oc ignition rhcct` or something, see also https://github.com/openshift/oc/pull/628

IOW wherever we say `fcct` we can say `oc ignition fcct` for OKD/FCOS too or so.

Alternatively we could add this functionality directly to MachineConfig, something like an even simpler
```
storage:
  boot_raid:
    - /dev/vda
    - /dev/vdb
```

Basically this path is making MachineConfig a peer of FCC.

(Or we could just punt this section for a second phase)